components/inputs/input-date.js

```diff
@@ -297,25 +333,34 @@ class InputDate extends LocalizeCoreElement(LitElement) {
 		}
 	}
 
-	_handleSetToToday() {
+	async _handleSetToToday() {
 		const date = getToday();
-		this._updateValueDispatchEvent(formatDateInISO(date));
+		await this._updateValueDispatchEvent(formatDateInISO(date));
 		this._dropdown.close();
 		this.focus();
 	}
 
 	_setFormattedValue() {
-		this._formattedValue = this.value ? formatISODateInUserCalDescriptor(this.value) : (this.emptyText ? this.emptyText : '');
+		this._formattedValue = this._shownValue ? formatISODateInUserCalDescriptor(this._shownValue) : (this.emptyText ? this.emptyText : '');
 	}
 
-	_updateValueDispatchEvent(dateInISO) {
-		if (dateInISO === this.value) return;
+	async _updateValueDispatchEvent(dateInISO) {
+		if (dateInISO === this._shownValue) return; // prevent validation from happening multiple times for same change
+		this._shownValue = dateInISO;
+		this.validationCustomValid = !dateInISO || isDateInRange(getDateFromISODate(dateInISO), getDateFromISODate(this.minValue), getDateFromISODate(this.maxValue));
```

Comment:
Since this validation check is synchronous I would recommend using `setValidity` which is modeled after Form Associated Custom elements / the HTML 5 validation API.

It would look something along the lines of:
```
if (this.minValue && date.getTime() < getDateFromISODate(this.minValue).getTime()) {
    this.setValidity({ rangeUnderflow: true });
} else if (this.maxValue && date.getTime() > getDateFromISODate(this.maxValue).getTime()) {
    this.setValidity({ rangeOverflow: true });
} else {
    this.setValidity({});
}
```

Then to customize the messages you can override these:
```
get validationMessageRangeOverflow() {
    return 'It is too large, must be below max!';
}

get validationMessageRangeUnderflow() {
    return 'It is too low, must be above min!';
}
```

Reply:
Awesome, that seems to work well - thanks!